### DISCUSSION TITLE
XAML Binding Failures - False Positives?

### DISCUSSION THREAD

**Discussion by Cywizz** [dotnet/maui#5511] [Q&A]
Hi all,

In the process wrapping my head around everything, but I found something strange. I am not sure if it is me or a possible bug.

I use the MauiAppBuilder to inject my views and viewmodels for a navigation service I have. 
I also set the BindingContext property on the view's codebehind to the viewmodel I injected, which is presented to me via the view's constructor. The build in DI provided by maui is great, btw!

Because I do not set the binding context in the xaml declarations, the editor shows that it can't find the specified property, which is expected at design time. It works fine for normal properties at run time, and does not show any binding failures. Great!

However, I have an ObservableCollection property that are bound to a listview's ItemSource, and within the DataTemplate of the listview, I individually bind to the relevant properties of the type in the collection.
When I run the app, it shows binding errors in the XAML Bind Failures window, however everything is bound and is displaying correctly in the listview!?
I do not mind seeing the errors, provided that it works, but I am concerned that something might bite me later that I am not aware of.
Or could it be a false positive by the framework due to the way I inject and use the views/models?

**Comment by baswelsink** [comment]
Same issue here, any idea yet?

**Reply by showtroylove** [reply]
I'm also seeing this with the simplest UI.  VM ->  Observerable Property -> ListView -> Observable Property Item -> Property.  The XAML Binding Failures window reports the Data Context (where it's trying to find the property) as the VM.    Path it should be looking is VM->Observable[i]->Property.   The page renders, but it appears to be at the expense of performance.  Conceivably it's performing visual tree hierarchy reflection to find the values.

Tried using x:DataType on the \<DataTemplate x:DataType="x:Type"\> string and x:Type form.  No Joy!

@Cywizz are you using nested layouts (e.g. Grids, etc.)?  My suspicions is nesting.

**Reply by sidewinder94** [reply]
If this is irking you too much, I've found a workaround, I don't like it too much, but I prefer that instead of being clobbered by "normal" errors, which make it too easy to miss real binding errors.

In each of your bindings, define the source as a follow: `Source={RelativeSource Mode=FindAncestorBindingContext, AncestorType={x:Type vm:YourType}}`.

I don't have any idea as to why this is necessary since things seem to work at runtime even though errors are reported. I'm also using this on a same scale project, so no ideas if this would impact performance in any way.

**Reply by showtroylove** [reply]
I’m already using that as my workaround. Although I use the reference form of this binding, it seems to be a .NET MAUI bug in the way it resolves bindings. Interestingly, other XAML implementation frameworks (such as WPF and WinUI) don’t seem to complain when using similar bindings.

**Comment by DanTravison** [comment]
I'm seeing the same issue when I use ListView.ItemTemplate.  I see a Diagnostics.BindingDiagnostics: Warning for each item indicating the property is not found on the page's ViewModel. I'm binding the ItemSource to the view model's Items property and have tried setting x:DataType on the <DataTemplate> element, the <ViewCell> element and, lastly, on the child element under the ViewCell.  I get the warning regardless of which element I set in the DataTemplate.

I've also tried removing the x:DataType from all elements in the page with no success.

From the behavior, it appears that x:DataType is completely ignored on items added via ItemsSource.

I have three cases where I'm using ListView with ItemsSource.  In one, I use a direct DataTemplate for the ItemTemplate. In the two others, I use a DataTemplateSelector.  In all cases, I see this warning.

The problem is exacerbated in one of the usages because the items are also collections and the DataTemplate uses a BindableLayout to render them.  In that case, I get the warning as well as an InvalidCastException attempting to cast the viewmodel to an IEnumerable instead of the contained item.

Since the bindings end up working at runtime, it appears to be failing and then succeeding (i.e., it's doing twice the work).

**Reply by DanTravison** [reply]
Note the document on compiled bindings states that x:DataType is needed on the DataTemplate. In fact, is does not work in any of the ListView usages I've tried.

see: https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/compiled-bindings?view=net-maui-7.0#use-compiled-bindings-in-a-datatemplate

**Comment by DanTravison** [comment]
I've dug into this a little more and found the culprit: 
The ListView has an ItemsSource and ItemTemplate defined. The ItemTemplate contains a StackLayout which also has an ItemsSource and ItemTemplate defined.

The problem occurs when the calling Element.SetParent on the ViewCell. This calls BindableObject.SetInheritedBindingContext which ends up setting the ListView's BindingContext on the ViewCell instead of the intended value from the ItemsSource. This propogates down the the StackLayout's ItemsSource which expects an IEnumerable.  Since the ListView's BindingContext is not an IEnumerable, an InvalidCastException(Object must implement IConvertible) is raised by Convert.ChangeType.  If the ListView's BindingContext does implement IEnumerable, a flood of BindingDiagnostics: Warning occur for each Cell also due to the incorrect value.

I see this sequence occur twice for each cell as the ListView is populated causing a user noticeable delay in displaying the associated page.

To me, this is clearly a bug.  If an ItemsSource is defined on the ListView, the cell should be getting the associated object in the ItemsSource for its BindingContext, NOT the parent's BindingContext.

**Comment by showtroylove** [comment]
This issue appears to have worked it way to the right people at .NET MAUI central.  I no longer experience these binding issues.  Most likely a bug that was resolved, although I haven't confirmed.

**Reply by allyoucan** [reply]
Nope. The bug is still there

**Reply by COMH-AIQ** [reply]
I'm hoping this gets fixed as it sounds exactly like the situation I'm in. The UI is still displaying the text, but I'm getting the binding diagnostics warnings about missing properties. As of today (Feb. 9, 2024) it is still a problem for me.

**Comment by MarcelWolterbeek** [comment]
Workaround to prevent the warnings (and InvalidCastExceptions):
In your ViewCell implementation add this code:
```
protected override void OnBindingContextChanged()
{
    if (BindingContext is TheViewModelForYourListItem)
        base.OnBindingContextChanged();
}
```
